File: Sample/src/main/java/com/brightyu/androidcommon/modules/account/password/forgot/ForgotPasswordActivity.java
/**
 * Copyright (C) 2016 The yuhaiyang Android Source Project
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.brightyu.androidcommon.modules.account.password.forgot;

import android.os.Bundle;
import android.view.View;

import com.bright.common.widget.TopBar;
import com.bright.common.widget.VerifyCodeButton;
import com.bright.common.widget.edittext.EditTextPro;
import com.bright.common.widget.loading.LoadingDialog;
import com.brightyu.androidcommon.R;
import com.brightyu.androidcommon.modules.base.AppBaseActivity;

/**
 * 修改密码和重置密码一系类的东西
 * 和注册分开预防后期业务更改
 */
public class ForgotPasswordActivity extends AppBaseActivity implements View.OnClickListener, ForgotPasswordContract.View {

    private EditTextPro mInputPhone;
    private EditTextPro mInputVerify;
    private EditTextPro mInputPassword;
    private EditTextPro mInputPasswordEnsure;

    private VerifyCodeButton mVerifyCodeButton;

    private ForgotPasswordContract.Presenter mPresenter;

    private LoadingDialog mLoadingDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_password);
        mPresenter = new ForgotPasswordPresenter(this);
    }

    @Override
    protected void initViews() {
        super.initViews();
        TopBar topBar = (TopBar) findViewById(R.id.top_bar);
        topBar.setOnTopBarListener(this);

        mInputPhone = (EditTextPro) findViewById(R.id.phone);
        mInputVerify = (EditTextPro) findViewById(R.id.verify_code);
        mInputPassword = (EditTextPro) findViewById(R.id.password);
        mInputPasswordEnsure = (EditTextPro) findViewById(R.id.ensure_password);

        mVerifyCodeButton = (VerifyCodeButton) findViewById(R.id.send_verify_code);
        mVerifyCodeButton.setOnClickListener(this);

        View register = findViewById(R.id.submit);
        register.setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.send_verify_code:
                mVerifyCodeButton.showLoading();
                mPresenter.sendVerifiyCode();
                break;
            case R.id.submit:
                String phone = mInputPhone.getInputText();
                String verifyCode = mInputVerify.getInputText();
                String password = mInputPassword.getInputText();
                String passwordEnsure = mInputPasswordEnsure.getInputText();
                mPresenter.resetPassword(this, phone, verifyCode, password, passwordEnsure);
                break;
        }
    }


    @Override
    public void showSendVerifySuccess() {
        mVerifyCodeButton.startTiming();
    }

    @Override
    public void showSendVerifyFail(String message) {
        mVerifyCodeButton.reset();
        dialog(message);
    }

    @Override
    public void showLoading(String message, boolean dialog) {
        mLoadingDialog = LoadingDialog.show(this, mLoadingDialog);
    }

    @Override
    public void dismissLoading(boolean dialog) {
        LoadingDialog.dismiss(mLoadingDialog);
    }

    @Override
    public void showError(String message, boolean dialog, int errorType) {
        dialog(message);
    }

    @Override
    public void showSuccess(String message) {

    }
}

File: app/src/main/java/com/ishow/noah/modules/account/password/forgot/ForgotPasswordActivity.kt
/**
 * Copyright (C) 2016 The yuhaiyang Android Source Project
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ishow.noah.modules.account.password.forgot

import android.content.Intent
import android.os.Bundle
import android.view.View
import com.ishow.common.extensions.toast
import com.ishow.common.utils.databinding.bus.Event
import com.ishow.common.utils.router.AppRouter
import com.ishow.common.utils.watcher.EnableTextWatcher
import com.ishow.common.utils.watcher.VerifyCodeTextWatcher
import com.ishow.common.utils.watcher.checker.PhoneNumberChecker
import com.ishow.noah.R
import com.ishow.noah.databinding.ActivityPasswordBinding
import com.ishow.noah.modules.account.login.LoginActivity
import com.ishow.noah.modules.base.mvvm.view.AppBindActivity
import com.ishow.noah.utils.checker.PasswordChecker

/**
 * Created by yuhaiyang on 2018/8/8.
 * 修改密码和重置密码一系类的东西
 * 和注册分开预防后期业务更改
 */
class ForgotPasswordActivity : AppBindActivity<ActivityPasswordBinding, ForgotPasswordViewModel>() {

    private lateinit var mVerifyCodeWatcher: VerifyCodeTextWatcher
    private lateinit var mSubmitWatcher: EnableTextWatcher

    private lateinit var mViewModel: ForgotPasswordViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        bindContentView(R.layout.activity_password)
    }

    override fun initViewModel(vm: ForgotPasswordViewModel) {
        super.initViewModel(vm)
        observeLiveData(vm)
        mViewModel = vm
    }

    override fun initViews() {
        super.initViews()
        mVerifyCodeWatcher = VerifyCodeTextWatcher()
            .setEnableView(binding.sendVerifyCode)
            .addChecker(binding.phone, PhoneNumberChecker())

        mSubmitWatcher = EnableTextWatcher()
            .setEnableView(binding.submit)
            .addChecker(binding.phone, PhoneNumberChecker())
            .addChecker(binding.verifyCode)
            .addChecker(binding.password, PasswordChecker(context))
            .addChecker(binding.ensurePassword, PasswordChecker(context))
    }

    fun onViewClick(v: View) {
        when (v.id) {
            R.id.sendVerifyCode -> {
                binding.sendVerifyCode.showLoading()
                mViewModel.sendVerifyCode(binding.phone.inputText)
            }
            R.id.submit -> {
                mViewModel.resetPassword(
                    binding.phone.inputText,
                    binding.verifyCode.inputText,
                    binding.password.inputText,
                    binding.ensurePassword.inputText
                )
            }
        }
    }

    private fun observeLiveData(vm: ForgotPasswordViewModel) = vm.run {
        verifyCodeStatus.observe(activity, { onVerifyCodeStatusChanged(it) })
        resetState.observe(activity, { resetSuccess() })
    }

    private fun onVerifyCodeStatusChanged(status: Event<Boolean>) {
        status.getContent()?.let { success ->
            if (success) {
                binding.sendVerifyCode.startTiming()
            } else {
                binding.sendVerifyCode.reset()
            }
        }
    }

    private fun resetSuccess() {
        toast(R.string.reset_password_success)
        AppRouter.with(context)
            .target(LoginActivity::class.java)
            .flag(Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK)
            .finishSelf()
            .start()
    }
}
